text/0000-knowntypeid.md

```diff
@@ -0,0 +1,140 @@
+- Feature Name: knowntypeid
+- Start Date: 2019-08-01
+- RFC PR: [rust-lang/rfcs#2738](https://github.com/rust-lang/rfcs/pull/2738)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a new unsafe trait `KnownTypeId` to `core::any`, and implement it for all types. This has a single method `type_id` with the signature `fn type_id(&self) -> TypeId where Self: 'static`. Traits wanting to support downcasting can add it as a supertrait, without breaking backwards compatibility.
+
+# Motivation
+[motivation]: #motivation
+
+This enables traits other than `core::any::Any`, such as `std::error::Error`, to soundly support downcasting in safe code. The initial implementation of this relied on simply adding a method to `Error`, but this was found to be unsound to override.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The new `core::any::KnownTypeId` unsafe trait provides a similar functionality to `core::any::Any`, providing a `type_id` method for types where `Self: 'static`. However, the trait itself does not impose any requirements on its implementors. This is unlike `core::any::Any`, which requires `Self: 'static` for the trait to be implemented. It is also implemented for every type. These features enable it to be added as a supertrait without breaking backwards compatibility. The trait's definition follows:
+
+```rust
+pub unsafe trait KnownTypeId {
+    fn type_id(&self) -> TypeId where Self: 'static {
+        TypeId::of::<Self>()
+    }
+}
+
+unsafe impl<T: ?Sized> KnownTypeId for T {}
```

Comment:
Maybe add a comment explaining that the only purpose of this instance is to prevent people from instantiating this trait for their own types.

AFAIK this is good enough even with specialization as the function is not marked `default`, but I am not sure.

Reply:
> AFAIK this is good enough even with specialization as the function is not marked `default`, but I am not sure.

The fact that it is an `unsafe trait` should make it enough even in the presence of specialization and even with `default` fn because the unsoundness would be the fault of the person doing the overriding. (But I don't see why we'd ever put `default` on this to allow specialization on it anyways...)

Reply:
The fact that we have to use two mechanisms to be extra sure about this does not instill confidence though. ;)

With the `unsafe` argument, this wildcard impl is technically unnecessary.

Reply:
> The fact that we have to use two mechanisms to be extra sure about this does not instill confidence though. ;)

Wait... I would think that extra safeguards would inspire more confidence...? 😂

Reply:
To me it seems like "yeah we *think* this is enough but we are not actually sure so let's add another layer". But maybe that's not a bad idea.

I just don't like redundancy.^^

Reply:
We could say that the blanket impl is for convenience and the unsafe marking makes it safe even in the presense of specialization (whatever form that takes)

Reply:
That was the intention, yes.

Reply:
> I just don't like redundancy.^^

Redundancy is in fact key to the soundness / structural integrity of any building; ask an architect. ;)

Reply:
But come on, what to architects know about this? Just compare the stuff they did and how well it stands the test of time, vs the rock-solid foundations that we software people build.

...

Okay well I guess you got a point. Still I prefer proofs over redundancy. ;)

Reply:
@leo60228 if there is a redundant-looking impl block with no body, there should be a comment saying why. That's all I ask. ;)

Reply:
@RalfJung Done.

Reply:
@gnzlbg The trait should be implemented for every type anyway. I don't see any reason to seal it when it already has a blanket implementation (so you'd have to use specialization) and is unsafe.